olp-cpp-sdk-core/src/http/winhttp/NetworkWinHttp.cpp

```diff
@@ -0,0 +1,1006 @@
+                                      proxy.GetUsername().end());
+  proxy_string_stream << ':';
+  proxy_string_stream << proxy.GetPort();
+
+  return proxy_string_stream.str();
+}
+
+}  // namespace
+
+namespace olp {
+namespace http {
+
+#define LOGTAG "WinHttp"
+
+NetworkWinHttp::NetworkWinHttp()
+    : http_session_(NULL),
+      thread_(INVALID_HANDLE_VALUE),
+      event_(INVALID_HANDLE_VALUE) {
+  request_id_counter_.store(
+      static_cast<RequestId>(RequestIdConstants::RequestIdMin));
+
+  http_session_ = WinHttpOpen(L"OLP Http Client", WINHTTP_ACCESS_TYPE_NO_PROXY,
+                              WINHTTP_NO_PROXY_NAME, WINHTTP_NO_PROXY_BYPASS,
+                              WINHTTP_FLAG_ASYNC);
+
+  if (!http_session_) {
+    EDGE_SDK_LOG_ERROR(LOGTAG, "WinHttpOpen failed " << GetLastError());
+    return;
+  }
+
+  WinHttpSetStatusCallback(
+      http_session_, (WINHTTP_STATUS_CALLBACK)&NetworkWinHttp::RequestCallback,
+      WINHTTP_CALLBACK_FLAG_ALL_COMPLETIONS | WINHTTP_CALLBACK_FLAG_HANDLES, 0);
+
+  event_ = CreateEvent(NULL, TRUE, FALSE, NULL);
+
+  // Store the memory tracking state during initialization so that it can be
+  // used by the thread.
+  thread_ = CreateThread(NULL, 0, NetworkWinHttp::Run, this, 0, NULL);
+  SetThreadPriority(thread_, THREAD_PRIORITY_ABOVE_NORMAL);
+}
+
+NetworkWinHttp::~NetworkWinHttp() {
+  std::vector<std::shared_ptr<ResultData>> pendingResults;
+  {
+    std::unique_lock<std::recursive_mutex> lock(mutex_);
+    std::vector<std::shared_ptr<RequestData>> requestsToCancel;
+    for (auto& request : http_requests_) {
+      requestsToCancel.push_back(request.second);
+    }
+    while (!requestsToCancel.empty()) {
+      std::shared_ptr<RequestData> request = requestsToCancel.front();
+      WinHttpCloseHandle(request->http_request);
+      request->http_request = NULL;
+      pendingResults.push_back(request->result_data);
+      requestsToCancel.erase(requestsToCancel.begin());
+    }
+  }
+
+  if (http_session_) {
+    WinHttpCloseHandle(http_session_);
+    http_session_ = NULL;
+  }
+
+  if (event_ != INVALID_HANDLE_VALUE) SetEvent(event_);
+  if (thread_ != INVALID_HANDLE_VALUE) {
+    if (GetCurrentThreadId() != GetThreadId(thread_)) {
+      WaitForSingleObject(thread_, INFINITE);
+    }
+  }
+  CloseHandle(event_);
+  CloseHandle(thread_);
+  thread_ = event_ = INVALID_HANDLE_VALUE;
+
+  {
+    std::unique_lock<std::recursive_mutex> lock(mutex_);
+    http_connections_.clear();
+    while (!results_.empty()) {
+      pendingResults.push_back(results_.front());
+      results_.pop();
+    }
+  }
+
+  for (auto& result : pendingResults) {
+    std::unique_lock<std::recursive_mutex> lock(mutex_);
+    if (result->user_callback) {
+      result->user_callback(
+          NetworkResponse()
+              .WithRequestId(result->request_id)
+              .WithStatus(static_cast<int>(ErrorCode::OFFLINE_ERROR))
+              .WithError("Offline"));
```

Comment:
```suggestion
              .WithError("Offline: network is deinitialized"));
```
To align with Linux implementation. Reason - to distinguish this OFFLINE error from regular offline error message